blog/coap/coap.md

```diff
@@ -0,0 +1,245 @@
+---
+slug: supermq-coap-adapter
+title: "SuperMQ CoAP Adapter: Building IoT Communication with Constrained Devices"
+authors: steve
+description: "A comprehensive guide to using SuperMQ's CoAP adapter for lightweight IoT communication - from device authentication to real-time data streaming with constrained devices."
+tags:
+  [
+    iot,
+    coap,
+    constrained-devices,
+    udp,
+    golang,
+    real-time,
+    senml,
+    message-broker,
+    supermq
+  ]
+---
+
+# SuperMQ CoAP Adapter: Building IoT Communication with Constrained Devices
+
+## Introduction
+
+IoT has changed how we connect to the physical world. From powerful edge computers to tiny battery-powered sensors, billions of devices now communicate with each other and the cloud. But here's the thing - HTTP, which works great for websites, is often too heavy for small IoT devices with limited power and bandwidth.
+
+That's where CoAP comes in. CoAP (Constrained Application Protocol) was built specifically for these resource-constrained devices. It runs on UDP instead of TCP, making it much lighter than HTTP, while still keeping the useful REST-style features developers love. SuperMQ's CoAP adapter connects these small devices to the larger SuperMQ platform without breaking a sweat.
+
+SuperMQ doesn't put all its eggs in one basket when it comes to protocols. You can use HTTP for web apps, MQTT for publish-subscribe messaging, WebSocket for real-time web communication, and CoAP for constrained devices. This means you can pick the right tool for each job while everything still plays nicely together.
+
+This post walks through how SuperMQ's CoAP adapter works under the hood and shows you real examples of using it in your IoT projects.
+
+## System Architecture
+
+SuperMQ's CoAP adapter is built tough - it scales well, keeps things secure, and doesn't fall over when things get busy. Let's break down what makes it tick:
+
+### Core Technologies
+
+**Go-CoAP Library**: We use the [plgd-dev/go-coap](https://github.com/plgd-dev/go-coap) library as our foundation. It's fast and handles all the nitty-gritty CoAP details:
+- Managing UDP messages and making sure they're reliable
+- Parsing and checking CoAP options
+- Implementing the Observe pattern for real-time updates
+- Handling different content formats
+
+**gRPC Integration**: For talking to other SuperMQ services, we use gRPC. It's type-safe and fast:
+- **Clients Service**: Takes care of authentication and managing clients
+- **Channels Service**: Handles permissions and channel stuff
+- **Message Broker**: Routes messages through SuperMQ
+
+**Message Broker Integration**: SuperMQ works with different message brokers (NATS, RabbitMQ) through a common interface, so messages get delivered reliably and pub/sub works smoothly.
```

Comment:
This is very vague, May be we can refer to existing blog section, where it explains how the message from adapter is published to NATS